solver/utils.go

```diff
@@ -95,3 +96,9 @@ func eventFromArg(arg crud.Arg) diff.Event {
 	}
 	return event
 }
+
+func isDocument(obj interface{}) bool {
+	a := reflect.TypeOf(obj)
+	b := reflect.TypeOf(&state.Document{})
+	return a == b
```

Comment:
Would simple type assertion not work?

```go
func isDocument(obj interface{}) bool {
	_, ok := obj.(*state.Document)
	return ok
}
```

Reply:
Or we could even do this inline in the `if` which calls `isDocument` today:

```go
      if oldObj, ok := e.OldObj.(*state.Document); ok {
        diffString, err = getDocumentDiff(oldObj, e.Obj.(*state.Document))
      } else {
        diffString, err = getDiff(e.OldObj, e.Obj)
      }
```

without the need to define a function and write a unit test.

Reply:
It doesn't need a function; we broke it out specifically because we [wanted to unit test that part of the code](https://github.com/Kong/deck/pull/397#discussion_r636492718), and the calling function lacks its own test.

Reply:
The problem we're solving is *exactly* an instance of the pattern below:

```
if x is of type A
  make a type assertion obvious from the condition above, and do something with x
else
  do something else
```

By hiding "x is of type A" we're making the operation more complex:
```
if (x is a document, whatever this means)
  make an unchecked type assertion not obviously valid from the
    condition above (without looking into the implementation details of the function), 
    and do something with x
else
  do something else
```

Also we have a unit test that verifies that:
1. a type assertion in Go works as intended
2. that the abstraction implements a type assertion

There is no point in testing (1). We don't test the plus operator by checking that `2+2 == 4`; we don't test the `if` expression by checking that if the condition is false, the block won't get executed. Why would we check that a type assertion works as expected?

Testing (2) would make sense if the abstraction provided value. I think it does not. It introduces some concept of "an untyped variable being _a document_", and leads to a type assertion being made in the conditional block that is not checked for validity locally. This will make the life of a person debugging this in the future harder.

Reply:
Also, with the current implementation, the type assertion happens twice (first in `isDocument` and then in the conditional block). After the proposed change gets made, it will happen only once (in the condition).